### ISSUE TITLE
Helm doesn't updated resource though Chart's template changed

### ISSUE THREAD

**Issue by whitecolor** [helm/helm#1873]
Client version "v2.1.3"

I'm using customized `nginx-lego` chart from a folder. 

First I installed it as it is, then adde arguments here:
https://github.com/kubernetes/charts/blob/master/stable/nginx-lego/templates/nginx-deployment.yaml#L38

So in charts templates I have changed arguments:
![image](https://cloud.githubusercontent.com/assets/736697/22290925/66e5947e-e325-11e6-8aeb-885c8d6cdbf2.png)

Then I do `helm upgrade` release, the deployment is not updated there is still two arguments:
![image](https://cloud.githubusercontent.com/assets/736697/22290896/3b7e5046-e325-11e6-9897-bdb9c2b37494.png)

Is there a way to imperatively upgrade the deployment?

**Comment by technosophos** [comment]
Can you confirm for me whether this is a duplicate of #1844 ?

**Comment by whitecolor** [comment]
Well probably I can not be sure, I just say that I have updated args of `container` in `deployment` template, and helm doesn't update deployment while `upgrade` operation, that is all I can confirm.

**Comment by technosophos** [comment]
Is this fixed in 2.2.0?

**Comment by whitecolor** [comment]
Havent tried this yet with 2.2.0, will try.

**Comment by whitecolor** [comment]
@technosophos @thomastaylor312 
The same is here on the latest version

I changed container `args` in deployment (`--sub-host lt` from ``--sub-host`, `lt`)
![image](https://user-images.githubusercontent.com/736697/26875972-961505f0-4b9d-11e7-80ec-7a16af121c2d.png)

Corresponding pod is recreated but has old `args` config:

![image](https://user-images.githubusercontent.com/736697/26875987-a8ff01c0-4b9d-11e7-9077-94f011f11af9.png)

**Comment by whitecolor** [comment]
Still see this on 2.4.3

When I update args on deployment templte container, replica set and created pod are not getting this update.

Have you checked this bug? Should I try to update to 2.5?

**Comment by whitecolor** [comment]
@technosophos has heml team looked into it?

**Comment by fuhbar** [comment]
Can confirm this still exists in Helm 2.5.0 and k8s 1.6.6. 

Updating args or even annotations in the values file will lead to correct user supplied and computed values, but the manifest of the `Deployment` gets not updated.

**Comment by whitecolor** [comment]
Well, I'm seeing the situation that `Deployment` actually is updated and has new `args` but underlying replicasets and pods hold old values.

**Comment by thomastaylor312** [comment]
@whitecolor Are they orphaned pods with the new pods around also or are there no new pods at all?

**Comment by whitecolor** [comment]
They was old pods. I changed args in deployment now (using heml ), And they were added properly to pods now. I don't know we still see happening sometimes.

**Comment by alonl** [comment]
We encounter a similar issue as well. We deploy an umbrella chart using `helm upgrade`. I run the command with `--debug` so I see the correct updated yaml of the charts being printed, but the `Deployment` (not the `ReplicaSet`) isn't updated. In Tiller logs, it seems like that it doesn't even recognize the changes: ```2017/07/31 11:50:51 Looks like there are no changes for Deployment ...```

**Comment by xforze** [comment]
Any news on this ?
I notice the same issue "Looks like there are no changes for Deployment" but debug output from helm definitively differs from current deployment.
Version: 2.7.0

**Comment by thomastaylor312** [comment]
@xforze Do you have a chart we can replicate this with?

**Comment by bacongobbler** [comment]
ping

**Comment by valer-cara** [comment]
Encountered similar situation on 2.6.1: Changing args in a deployment were not reflected in the cluster.

I've also tried deleting the deployment and doing a `helm upgrade`. Tiller logs "Created new deployment", however this one has the old args as well (which is the most confusing part of this).

**Comment by whitecolor** [comment]
@technosophos 
I believe helm team should investigate carefully code that effects updating of the argument list at least.

**Comment by valer-cara** [comment]
I've manged to solve the problem with `helm --reset-values foobar ./foobar`.

Perhaps the issue happened because in the past I ran it with `helm upgrade foobar -f ./foobar/values.yaml ./foobar`? I see that `--reuse-values` is false by default so I expect that not to be the case.

**Comment by xforze** [comment]
@thomastaylor312 

I attached the chart files which we use. The Problem changed meanwhile :-)
If we try to deploy a newer version of our application, a new Replicaset will be installed instead of updating the existing one (every time I do an upgrade a new rs appears).

Im always using this command line for installation and upgrade
`/usr/local/bin/helm upgrade --install --wait --debug --kube-context default-system --namespace nginx-code-service --tiller-namespace gocdmaster-pocu-helper -f ./helm.yml -f values.yml nginx-code-service-dev helm-charts/spring-boot
`
For the 1st deploy of the application, I set the following values in values.yaml (<- is attached):
- tag: "1.12"
- version:  "3.4.7"
and run the helm command. This will install the nginx image 1.12 in a Replicaset

For the 2nd deploy (upgrade) I set the following values in values.yaml:
- tag: "1.13"
- version:  "3.4.9"
and run the helm command again. This will install a new (second) Replicaset with nginx image 1.13 instead of upgrading the existing one. 

Please let me know if you need further imformation.

Cheers, Thomas

[chart.tar.gz](https://github.com/kubernetes/helm/files/1450057/chart.tar.gz)

**Comment by thomastaylor312** [comment]
@xforze Thanks for putting that together. So I am wondering if the RS is getting orphaned because you are changing the labels and `matchSelector` with each change in the variables. That is in line with what I have seen before when changing the labels in a deployment like that

**Comment by thomastaylor312** [comment]
@valer-cara There were some issues with changes not being reflected on failed deploys that we just merged in (see #3097). Could that be the issue?

**Comment by valer-cara** [comment]
@thomastaylor312 - i can't tell for sure, but it sounds plausible. GG on that fix!

**Comment by thomastaylor312** [comment]
@valer-cara Let me know if the newer version fixes it for you

@xforze I never heard back on my last response to you. Are you still experiencing the same issue?

**Comment by jasquat** [comment]
We were able to reproduce this using kubernets 1.8.3 and helm 2.8.0 (and 2.7.2). We created a script to replicate it which is at https://github.com/jasquat/repro-helm-issue-1873.

**Comment by JrCs** [comment]
I have the same issue with helm 2.8.2 and kubernetes 1.9.3.
If i have deployed a deployment with helm with a bad image tag for my containers, when i try to fix the image tag and redeploy with helm the deployment is not upgrade (keeping the old image tag).
I use the helm options: upgrade --install --reset-values.

**Comment by JrCs** [comment]
I have the same issue with helm 2.8.2 and kubernetes 1.9.3.
If i have deployed a deployment with helm with a bad image tag for my containers, when i try to fix the image tag and redeploy with helm the deployment is not upgrade (keeping the old image tag).
I use the helm options: upgrade --install --reset-values.

**Comment by thomastaylor312** [comment]
Somehow this didn't get labelled as a bug after we got the reproducible example

**Comment by gianrubio** [comment]
I'm using similar steps from @jasquat to reproduce the issue, here are my scenario

### 1. Create a chart
```console
helm create error-chart
helm install --name error-chart error-chart # Rev 1

Giancarlos-MBPro:ddy_charts grubio$ helm ls
NAME            REVISION        UPDATED                         STATUS          CHART                   NAMESPACE
error-chart     1               Wed May  2 16:23:04 2018        DEPLOYED        error-chart-0.1.0       default
``` 

### 2. Force a deployment failure using a non existent image

```console
Giancarlos-MBPro:ddy_charts grubio$ helm upgrade  --wait --set image.repository=buuuu \
 --timeout 10 error-chart ./error-chart

Error: UPGRADE FAILED: timed out waiting for the condition
```

### 2.1 Verifying that deployment Image was correctly changed to buuuu
``` 
Giancarlos-MBPro:Downloads grubio$ k get deployment -n default -o yaml error-chart|grep image
      - image: buuuu:stable
        imagePullPolicy: IfNotPresent
```

### 3. Upgrade the chart, providing the right image

Giancarlos-MBPro:ddy_charts grubio$ helm upgrade  --set image.repository=nginx   error-chart ./error-chart

### 3.1 Deployments wasn't updated 😢 
``` 
Giancarlos-MBPro:Downloads grubio$ k get deployment -n default -o yaml error-chart|grep image
      - image: buuuu:stable
        imagePullPolicy: IfNotPresent
```


### Tiller logs during step 3
``` 
[kube] 2018/05/02 14:26:15 building resources from updated manifest
[kube] 2018/05/02 14:26:15 checking 2 resources for changes
[kube] 2018/05/02 14:26:15 Looks like there are no changes for Service "error-chart"
(THIS IS WRONG) =>> [kube] 2018/05/02 14:26:15 Looks like there are no changes for Deployment "error-chart" 
[tiller] 2018/05/02 14:26:15 executing 0 post-upgrade hooks for error-chart
```          

I investigate this issue and found in the code bellow that tiller compare the revision 3 with the configuration of the revision 1 (DEPLOYED). This isn't the right comparison, tiller should load the latest revision, in my scenario it's steps 2, and not the latest deployed revision. I'm willing to fix this issue but I'd like to discuss if what I said is correct.


https://github.com/kubernetes/helm/blob/0deb58ffdfa35f60a3706172e3bed4dd3a33bc65/pkg/storage/storage.go#L121-L156

**Comment by gianrubio** [comment]
> ... The reason things are not changing is because the diff to determine what changed is calculated between the last Helm release manifest and the re-rendered manifest generated by helm upgrade

@technosophos I just read [your comment](https://github.com/kubernetes/helm/issues/3685#issuecomment-373758896) above in another issue. Could you clarify if the last helm release manifest is the latest revision in `DEPLOYED`  state or the latest revision even if it's in `FAILED` status?

**Comment by estahn** [comment]
It seems we're running into this issue with the `stable/prometheus` chart.

The manifest shows `DEPLOYED`, the YAML output of `helm upgrade ...` shows the correct information (image change) but the image is not updated. But, a change in `replicaCount` is reflected.

![image](https://user-images.githubusercontent.com/362174/40696533-9265d940-6409-11e8-9faf-63de3f7e995c.png)

**Comment by wuestkamp** [comment]
Same issue here, has this been resolved?

**Comment by stefanthorpe** [comment]
Same issue here

**Comment by blakestoddard** [comment]
+1 to seeing if any progress has been made on this (tho it doesn't look like any has).

We're seeing this in the context of:

- Install a helm template that creates a deployment
- Manually scale the replica count via `kubectl scale`
- Run `helm upgrade` expecting Helm to reset the replica count back to what it's expecting
- No change occurs until the replica count is scaled manually back to the value that Helm is expecting (at which point it'll take control over the replica count once again)

Tiller is not convinced of any changes that need to be made even though the replica values differ between Kubernetes and the generated yaml/json:
```
[kube] 2018/09/17 14:27:07 Looks like there are no changes for Deployment "app"
```

(I also tested this without Helm in the mix and all is well -- if you `kubectl apply` a new deployment, manually change the replica count with `kubectl scale` and then `kubectl apply` the deployment config again, the replica count will be reset to what is in the applied yaml)

**Comment by Ehekatl** [comment]
same issue, whenever previous deploy is not in deployed state, the next helm upgrade will stop working.

**Comment by KIVagant** [comment]
I spend month preparing a project to migration to Helm. And today I did a deploy and found that labels were changed in a deployment but the image hash not. The related chart has a value 'imageVersion' and it has right version inside that was added with `sed` there during the chart build.

In my case, `values.yaml` has something like `imageVersion: "0012345"` and the template for Deployment has 
```
image: registry.mydomain.com/appname:{{ .Values.imageVersion }}
```

My deploy command looks like:

```
+ helm upgrade --dry-run --debug --wait --timeout 900 --install --set-string ciDeployBuildNumber=399 -f helm_values/values.yaml myproject charts/myproject/myproject-1.2.3.tgz
```

The result of the dry-run looks like:

```
apiVersion: apps/v1beta2
kind: Deployment
metadata:
  name: myproject
spec:
  selector:
    matchLabels:
      name: myproject
      chart: "myproject"
...
  template:
    metadata:
      labels:
        name: myproject
        updated: "2018-10-12"
        chart: "myproject"
        chart_version: "1.2.3"                         ← # The correct latest version
        ci_dep_build_number: "399"
    spec:
...
      containers:
      - name: myproject
        image: registry.myproject.com/myproject:0012345  # ← expected image hash
```

After the same command but without dry-run I see:

```
Release "myproject" has been upgraded. Happy Helming!
...

==> v1/Pod(related)

NAME                                             READY  STATUS       RESTARTS  AGE
myproject-6d955758bc-8nk7p                1/1    Running      0         14s
myproject-859d758df4-r5vp4                 1/1    Terminating  0         17h
```

And when I'm doing `kubectl get deployment myproject -o yaml |grep 'image:'` it's saying:
```
        image: registry.myproject.com/myproject:93abec9 # ← The wrong hash from the previous version
```

It's a disaster.

**Comment by KIVagant** [comment]
I did a test and yes, it's true. How to reproduce:

1. Deploy a Helm chart, for example, `myproject:1.1`
2. Run `kubectl set image deployment/myproject myproject=registry.myproject.com/myproject:oldversion`
3. Deploy a Helm chart again: `translations:1.1`

It will show that configuration has `image: registry.myproject.com/myproject:newversion`, and even will restart all pods. But inside the deployment's annotations the last applied command will be:
  ```annotations:
...
    kubernetes.io/change-cause: kubectl set image deployment/myproject myproject=registry.myproject.com/myproject:oldversion
      --record=true```

It's dangerous to use different set of commands for downgrade and upgrade-after-downgrade.

**Comment by bacongobbler** [comment]
Helm only diffs against the previously deployed chart and the chart to be installed, so the fact that the deployed resource was left unchanged is working as intended from Helm's point of view.

https://github.com/helm/helm/issues/3805 describes a potential long-term solution, but it won't work for existing releases because two-way merge patches and three-way merge patches are incompatible with eachother. Have a look at that thread for more detail.

**Comment by KIVagant** [comment]
Thank you, @bacongobbler , for the explanation. For users like me who don't understand Helm very deeply and just expect that "deployed" means "everything was updated" this behavior could be very surprising. But I understand that it's just my wrong expectations.

**Comment by nickdgriffin** [comment]
@bacongobbler I think there's a gap when it comes to Deployment resources though and failures caused by their ReplicaSets, for example:

1. Helm installs a chart successfully, everything is fine and it gets marked as "DEPLOYED" - the Deployment has maxUnavailable set to 50%
2. An upgrade is carried out which makes changes to both the Deployment to set maxUnavailable to 25% and the Pod template with a new image ID - this fails because of the Pods, causing Helm to time out waiting for the pods and the release to be marked as "FAILED"
3. Another upgrade is carried out, containing an update to the maxUnavailable to set it to 50% again and to the Pod template with another new image ID - but the Deployment resource never gets updated because of the comparison with revision 1 which matches

Doing "helm get" on revision 3 in the above example returns what one would expect the outcome to be, but the actual Deployment resource outside of the Pod template is actually at revision 2.

**Comment by bacongobbler** [comment]
I believe the issues laid out in this ticket may be resolved once #6124 is merged. #6124 takes live state into consideration during an upgrade or a rollback, which should solve many of the issues raised in this thread.

If I'm missing any cases, please let me know. Thanks! :)

**Comment by bacongobbler** [comment]
@nickdgriffin I think that's a separate issue altogether than the ones laid out earlier in this thread. Would you mind opening a separate ticket for that and we can label it correctly? Seems like a usage/documentation question more than a bug

**Comment by nickdgriffin** [comment]
@bacongobbler : I agree, I think it's the same case - I'll post over there, thanks!

**Comment by bacongobbler** [comment]
closed via #6124
